<a href="https://colab.research.google.com/github/gilandrade/Trilha-Data-Science/blob/master/1_2_NumPy_Pandas_gabarito.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# <center> Introdução ao NumPy e Pandas - Teoria e Exercícios</center>
___

## Conteúdo
1. [Introdução](#intro)
2. [Jupyter Básico](#basics)
3. [Pacotes (Packages)](#pacotes)<br>
4. [Encerramento](#encerramento)

<a id="intro"></a>
## 1. Introdução

Agora que você já teve uma pequena introdução ao mundo de Machine Learning, vamos te introduzir a uma das ferramentas mais importantes em análise de dados: a biblioteca Pandas, através da linguagem de programação Python. 

Utilizaremos o Jupyter Notebook para o acompanhamento da aula. Para isso, é pré-requisito sua instalação, bem como o Python 3. Ambos são disponibilizados se você instalar o software Anaconda 3. O formato da aula propõe que o aluno consiga acompanhar o material e fazer os exercícios ao mesmo tempo.

<a id="basics"></a>
## 2. Jupyter Básico

### 2.1. O que é Jupyter?
"Jupyter" é um acrônimo que significa Julia, Python, and R. Essas linguagens de programação foram as primeiras utilizadas na aplicação Jupyter, mas, atualmente, o notebook também suporta  [muitas outras linguagens](https://github.com/jupyter/jupyter/wiki/Jupyter-kernels).

Por se tratar de uma aplicação server-client, o Jupyter Notebook App permite ao usuário fazer a edição e execução de seus notebooks via web browser (Chrome, por exemplo). A aplicação pode ser executada em um PC sem acesso a Internet. 

- É possível executar um documento notebook step-by-step (uma célula por vez) pressionando shift + enter.
- É possível executar todo o notebook em um único passo. Para isso, basta clicar no menu Cell -> Run All.
- Para reiniciar o kernel (i.e., o engine), clique no menu Kernel -> Restart. Isso pode ser útil quando você quer reinicializar o processamentos desde o começo (nesse caso, as variáveis são deletadas, arquivos abertos são fechados, etc...).

**Keyboard Navigation**

Essa interface do Jupyter Notebook foi **otimizada para eficiente uso com o teclado**. Isso é possível porque há dois modos diferentes de atalhos com o teclado: modo de edição (edition mode) e modo de comando (command mode).

Os atalhos mais importantes são "Enter", que entra no modo de edição, e "Esc", que entra no modo de comando.

No modo de edição, a maior parte do teclado está disponível para digitar nas células do notebook. Por conta disso, no modo de edição não há muitos atalhos disponíveis. Já no modo de comando, o teclado inteiro pode ser utilizado para atalhos, portanto, existem muito mais atalhos. Se você clicar em *Help->Keyboard Shortcuts*, será possível ver toda a lista de atalhos disponível.

Recomenda-se aprender os atalhos do modo de comando na seguinte ordem:

- Navegação básica: enter, shift-enter, up/k, down/j
- Salvar o notebook: s
- Mudar o tipo da célula (code, markdown, etc...): y, m, 1-6, t
- Criação de célula: a, b
- Edição de célula: x, c, v, d, z
- Kernel operations: i, 0 (pressionar duas vezes)

Apesar de esses atalhos representarem o que você mais utilizará no cotidiano, existem algumas [dicas e macetes](https://www.dataquest.io/blog/jupyter-notebook-tips-tricks-shortcuts) que podem melhorar sua experiência com a aplicação Jupyter Notebook.

### 2.2 Execução de células

Não podemos aprender Data Science sem antes conhecer nossas ferramentas, né?

Clique na célula abaixo e pressione ctrl+Enter para executar o código.

In [ ]:
print("Hello World!") #Lembre-se que shift+enter vai executar a célula atual e selecionar a próxima célula

Hello World!


Parabéns!!! Você acabou de executar sua primeira linha de código no Jupyter. Foi bem fácil, né?

É assim que o notebook funciona, você pode executar o documento célula a célula e observar o que está sendo realizado.

Vamos, agora, acompanhar uma seção sobre os pacotes básicos mais utilizados para análise de dados em Python: o **Numpy** e o **Pandas**!

<a name="pacotes"></a>
## 3. Pacotes (Packages)

Um *package* é um diretório de scripts Python, também chamados de modules, com um objetivo em comum. Isso significa que cada script é um módulo que define funções, métodos e tipos. E esse módulos estão organizados em *packages*.

Quando você inicializa o Jupyter, apenas o package built-in é carregado. Para usar qualquer função, método ou objeto definido em outro módulo, você deve, primeiramente, importá-lo.

Vamos tentar fazer isso com o NumPy, um pacote que lida eficientemente com arrays e matrizes, que você certamente utilizará bastante em sua vida com o Python.

### 3.1 NumPy

NumPy é a abreviação de ‘Numerical Python’ ou ‘Numeric Python’. Ela é uma biblioteca *open-source* (ou seja, é um software cujo código original é disponibilizado livremente e pode ser distribuido e modificado) que oferece suporte à arrays e matrizes multidimensionais, provendo diversas funções matemáticas úteis em computação científica.

Mas por que você deveria utilizar o NumPy? É simples! As listas do Python funcionam como as arrays, no entanto, são lentas para utilização com grandes volumes de dados. No ramo de Ciência de Dados, velocidade e recursos durante o processamento são bem importantes! Dessa forma, o NumPy nos possibilita utilizar os objetos arrays, que são bem mais rápidos que as listas tradicionais do Python.

O **ndarray** é o objeto fundamental do NumPy. Este objeto é uma matriz N-dimensional, vamos entender melhor como este objeto funciona nas células abaixo.

#### 3.1.1 Importando o NumPy

In [ ]:
# Primeiramente, o que você acha que aconteceria se tentássemos utilizar algumas coisa de um pacote que ainda não foi carregado?
# Nós podemos testar isso com a função array, do módulo numpy, que retorna um array para uma lista dada
array([1, 2, 3])

NameError: name 'array' is not defined

Como esperado, tivemos um erro. Nós deveríamos carregar o pacote, então.
Para fazer isso, utilizamos a keyword 'import' seguida do nome do package.

In [ ]:
import numpy
numpy.array([1, 2, 3])

array([1, 2, 3])

Repare como precisamos especificar qual é o pacote ao qual a função ```array``` pertence. Poderia ficar muito chato se você tivesse que digitar os nomes dos pacotes todas as vezes que quisesse utilizar alguma coisa deles.

Para te ajudar com isso, o Python permite que você utilize um alias para o nome do pacote, através da keyword 'as', tornando seu uso um pouco mais prático, como mostrado a seguir.

In [ ]:
import numpy as np
np.array([1, 2, 3])

array([1, 2, 3])

Por último, é possível carregar apenas uma parte de um pacote com a expressão  ```from ... import ...```. Fazendo isso, você estará permitido a utilizar apenas a parte que você carregou do pacote, sem a necessidade de especificar o nome do pacote posteriormente. Veja.

In [ ]:
from numpy import array
array([1, 2, 3])

array([1, 2, 3])

Ótimo, você já sabe importar os pacotes que precisar!

Vamos observar o porquê o NumPy é tão poderoso e preferível às listas do Python.

Imagine que você tenha duas listas de dados de indivíduos, com alturas e pesos, e queira calcular o IMC (Índice de Massa Corporal) de cada um deles, como mostrado abaixo:

In [ ]:
altura = [1.81, 1.77, 1.69, 1.91]
peso = [89.0, 77.3, 55.9, 99.4]

# Calculando o IMC
imc = peso / altura ** 2

TypeError: unsupported operand type(s) for ** or pow(): 'list' and 'int'

Observe que o Python nos retornou um erro porque não é possivel realizar cálculos com listas, para isso vamos utilizar o NumPy!

In [ ]:
# Criando os arrays com o NumPy

np_altura = np.array(altura)
np_peso = np.array(peso)

# Calculando o IMC
imc = np_peso / np_altura ** 2
imc

array([27.16644791, 24.67362508, 19.57214383, 27.24706011])

O Numpy consegue realizar perfeitamente as operações elemento a elemento!

Mas preste bem atenção, pois algumas operações podem funcionar de forma diferente do que você imagina. Veja o exemplo abaixo:

In [ ]:
# Lista do Python
altura * 2

[1.81, 1.77, 1.69, 1.91, 1.81, 1.77, 1.69, 1.91]

In [ ]:
# NumPy array
np_altura * 2

array([3.62, 3.54, 3.38, 3.82])

#### 3.1.2 Selecionando subconjuntos de NumPy Arrays
A seleção de subconjuntos de NumPy arrays funciona de foram similar à listas de Python.

In [ ]:
# retornando o terceiro elemento da array
imc[2]

19.572143832498863

Você também pode selecionar subconjuntos baseados em condições, de forma que apenas os valores que satisfazem as condições serão retornados.

In [ ]:
imc[imc > 25]

array([27.16644791, 27.24706011])

Vamos fazer um exercício para fixação.

#### Exercício 3.1 (9 pontos)
*Correção Binária: O aluno recebe todos os pontos caso cumpra todos os objetivos do exercicio (independente da lógica ou estilo da programação), caso contrário, recebe zero.*

In [ ]:
# lista de pesos de castanhas
c_peso = [0.946, 0.918, 0.906, 0.904, 0.858, 0.774, 0.652, 0.516, 0.478, 0.404, 0.396, 0.364, 0.342, 0.304, 
            0.262, 0.208, 0.134, 0.974, 0.792, 0.792, 0.628, 0.552, 0.506, 0.478, 0.462, 0.436, 0.408, 0.378, 
            0.3, 0.298, 0.268, 0.252, 0.16, 0.114, 0.092, 0.936, 0.894, 0.744, 0.706, 0.694, 0.69, 0.652, 0.518, 
            0.508, 0.502, 0.5, 0.47, 0.44, 0.39, 0.384]

# Importe o numpy como np
import numpy as np

# Crie um numpy array a partir de c_peso: np_c_peso
np_c_peso=np.array(c_peso)

# Imprima o tipo de np_c_peso
print(type(np_c_peso))

<class 'numpy.ndarray'>


In [ ]:
c_kg_preco = 45.00

# Crie a partir do array de c_peso com a quantia gasta em cada compra: np_c_despesa
np_c_despesa=np_c_peso*c_kg_preco
np_c_despesa

array([42.57, 41.31, 40.77, 40.68, 38.61, 34.83, 29.34, 23.22, 21.51,
       18.18, 17.82, 16.38, 15.39, 13.68, 11.79,  9.36,  6.03, 43.83,
       35.64, 35.64, 28.26, 24.84, 22.77, 21.51, 20.79, 19.62, 18.36,
       17.01, 13.5 , 13.41, 12.06, 11.34,  7.2 ,  5.13,  4.14, 42.12,
       40.23, 33.48, 31.77, 31.23, 31.05, 29.34, 23.31, 22.86, 22.59,
       22.5 , 21.15, 19.8 , 17.55, 17.28])

In [ ]:
# Imprima o peso no indíce 20
print(np_c_peso[20])

0.628


#### 3.1.3  Array N-dimensional
Vamos verificar o tipo dos arrays criados acima!

In [ ]:
print(type(np_peso))

<class 'numpy.ndarray'>


**ndarrays** significam arrays N-dimensionais, vamos criar um numpy array multi-dimensional a partir de listas tradicionais do Python.

In [ ]:
np_2d = np.array([[1.81, 1.77, 1.69, 1.91],
                  [89.0, 77.3, 55.9, 99.4]])
np_2d

array([[ 1.81,  1.77,  1.69,  1.91],
       [89.  , 77.3 , 55.9 , 99.4 ]])

Cada sublista da lista corresponde à uma linha da array bi-dimensional criada.
Nós podemos verificar o tamanho da array usando o atributo "shape".

In [ ]:
np_2d.shape

(2, 4)

Podemos ver que o np_2d tem 2 linhas e 4 colunas.

Assim como a array unidimensional, também podemos selecionar um subconjunto de uma array multi-dimensional, usando o índice da linha e coluna como abaixo.
![Subsetting](https://imgur.com/08EIOjy.png)
Veja alguns exemplos abaixo.

In [ ]:
np_2d[0]

array([1.81, 1.77, 1.69, 1.91])

In [ ]:
# Selecionando a altura (primeira linha) do terceiro elemento
np_2d[0][2]

1.69

Basicamente nós selecionamos a linha, e a partir daquela linha fazemos outra seleção.

Também é possível selecionar utilizando vírgulas dentro de colchetes: array[linha, coluna]

In [ ]:
# Primeira linha e terceira colna
np_2d[0, 2]

1.69

In [ ]:
# Todas as linhas e segunda e terceira colunas
np_2d[:, 1:3] 

array([[ 1.77,  1.69],
       [77.3 , 55.9 ]])

#### Exercício 3.2 (9 Pontos)
*Correção Binária: O aluno recebe todos os pontos caso cumpra todos os objetivos do exercicio (independente da lógica ou estilo da programação), caso contrário, recebe zero.*

Abaixo temos uma lista de listas contendo informações de vendas de castanhas de uma loja. Cada lista representa uma venda que foi realizada. O primeiro elemento de cada lista é o dia que  venda foi feita, o segundo elemento representa o peso das castanhas compradas. Por fim, o terceiro elemento é a quantia paga pelas castanhas.

Com isso em mente, faça os seguintes exercícios:

In [ ]:
castanha = [[2, 0.946, 66.1], 
          [2, 0.918, 32.96], 
          [2, 0.906, 58.76],
          [2, 0.904, 29.14], 
          [2, 0.858, 59.96],
          [2, 0.774, 27.77],
          [2, 0.652, 42.3],
          [2, 0.516, 18.51], 
          [2, 0.478, 17.15],
          [2, 0.404, 28.22], 
          [2, 0.396, 7.88], 
          [2, 0.364, 7.24],
          [2, 0.342, 22.18], 
          [2, 0.304, 10.91], 
          [2, 0.262, 9.41], 
          [2, 0.208, 4.13],
          [2, 0.134, 9.36],
          [4, 0.974, 34.95],
          [4, 0.792, 51.38],
          [4, 0.792, 51.38], 
          [4, 0.628, 12.48], 
          [4, 0.552, 19.81], 
          [4, 0.506, 25], 
          [4, 0.478, 31], 
          [4, 0.462, 32.24],
          [4, 0.436, 28.28],
          [4, 0.408, 14.64],
          [4, 0.378, 13.56],
          [4, 0.3, 19.46],
          [4, 0.298, 10.69],
          [4, 0.268, 9.62],
          [4, 0.252, 16.34],
          [4, 0.16, 3.18],
          [4, 0.114, 4.09],
          [4, 0.092, 5.97],
          [5, 0.936, 65.33],
          [5, 0.894, 32.07],
          [5, 0.744, 48.28], 
          [5, 0.706, 25.34],
          [5, 0.694, 24.91], 
          [5, 0.69, 13.72], 
          [5, 0.652, 42.32], 
          [5, 0.518, 33.6], 
          [5, 0.508, 18.23],
          [5, 0.502, 35.09],
          [5, 0.5, 27.45], 
          [5, 0.47, 9.35], 
          [5, 0.44, 28.54],
          [5, 0.39, 7.76], 
          [5, 0.384, 21.08]]

# Crie um numpy array 2d a partir de castanha: np_castanha
np_castanha=np.array(castanha)

# Imprima o tipo de np_castanha
print(type(np_castanha))

# Imprima a forma de np_castanha
print(np_castanha.shape)

<class 'numpy.ndarray'>
(50, 3)


In [ ]:
# Selecione toda a segunda coluna de np_castanha: np_c_peso
np_c_peso=np_castanha[:,1]

# Imprima o preço da 14 venda
print(np_castanha[13,2])

# Imprima todas as vendas feitas após o dia 2
flag=-1
for counter in np_castanha[:,0]:
    flag=flag+1
    if counter != 2:
            print(np_castanha[flag,:])

10.91
[ 4.     0.974 34.95 ]
[ 4.     0.792 51.38 ]
[ 4.     0.792 51.38 ]
[ 4.     0.628 12.48 ]
[ 4.     0.552 19.81 ]
[ 4.     0.506 25.   ]
[ 4.     0.478 31.   ]
[ 4.     0.462 32.24 ]
[ 4.     0.436 28.28 ]
[ 4.     0.408 14.64 ]
[ 4.     0.378 13.56 ]
[ 4.    0.3  19.46]
[ 4.     0.298 10.69 ]
[4.    0.268 9.62 ]
[ 4.     0.252 16.34 ]
[4.   0.16 3.18]
[4.    0.114 4.09 ]
[4.    0.092 5.97 ]
[ 5.     0.936 65.33 ]
[ 5.     0.894 32.07 ]
[ 5.     0.744 48.28 ]
[ 5.     0.706 25.34 ]
[ 5.     0.694 24.91 ]
[ 5.    0.69 13.72]
[ 5.     0.652 42.32 ]
[ 5.     0.518 33.6  ]
[ 5.     0.508 18.23 ]
[ 5.     0.502 35.09 ]
[ 5.    0.5  27.45]
[5.   0.47 9.35]
[ 5.    0.44 28.54]
[5.   0.39 7.76]
[ 5.     0.384 21.08 ]


#### 3.1.4 Estatística básica com NumPy
Costumeiramente o primeiro passo para analisar nossos dados é conhecê-los através de estatística descritiva. O NumPy pode ser usado para obter essa visão inicial dos dados mesmo com grande quantidade de observações.

Vamos usar alguns atributos do NumPy para começar a analisar nossos dados.

In [ ]:
np_a_p = np.array([[1.81, 89.0],
                  [1.77, 77.3],
                  [1.69, 55.9],
                  [1.91, 99.4]])

# Calculando a media dos pesos
np.mean(np_a_p[:, 1])

80.4

In [ ]:
# calculando a mediana dos pesos
np.median(np_a_p[:, 1])

83.15

In [ ]:
# calculando os coeficientes de correlação entre pesos e alturas
np.corrcoef(np_a_p[:, 0], np_a_p[:, 1])

array([[1.        , 0.96933562],
       [0.96933562, 1.        ]])

In [ ]:
# calculando o desvio padrão dos pesos
np.std(np_a_p[:, 1])

16.16183776678878

In [ ]:
# calculando a soma dos pesos
np.sum(np_a_p[:, 1])

321.6

Alguns desses atributos já estão disponíveis no Python, no entanto, a principal diferença entre eles é a performance. Os atributos do NumPy são mais rápidos na execução do que os básicos do Python.

Lembrando que sempre que tiver dificuldade para entender algum atributo, você pode consultar a documentação do [Numpy](https://numpy.org/doc/).

Agora que aprendemos como o NumPy funciona, vamos aprender sobre uma das bibliotecas mais utilizadas para manipulação de dados em Python, o **Pandas**!

### 3.2 O que é o Pandas?

Pandas é uma biblioteca *open source*, que proporciona estruturas de dados e ferramentas de análise de dados de alta performance e fáceis de usar para Python. Vamos entender melhor o que alguns termos significam:
 - *open source*: assim como o NumPy, seu código original é disponibilizado livremente e pode ser distribuido e modificado. Isso significa que qualquer um pode contribuir para a evolução do Pandas!
 - alta performance: Pandas é escrito em Python, Cython e C. Isso permite que os cientistas de dados consigam o utilizar para lidar com conjuntos de dados muito grandes (daqueles que o excel não conseguiria nem abrir) e fazer operações sobre esses dados com facilidade. Dessa forma, Pandas torna nosso trabalho melhor e mais fácil provendo ótima performânce.
 - estruturas de dados e análises de dados: o motivo pelo qual Pandas existe. Muitas vezes precisamos obter algo com significado de dados crus como documentos de textos, tabelas e etc. Pandas é capaz de lidar com esses tipos de dados para que possamos analisá-los.
Em resumo, **Pandas fornece estrutura de dados especializadas e ferramentas para manipulação de dados**. Sua ótima performânce, facilidade de uso e comunidade dedicada são as principais razões de sua vasta adoção entre cientistas de dados. Vamos utilizá-lo!
Se você está utilizando o Anaconda, já deve ter ele instalado. Senão, é possível achar os passos de instalação [neste link](https://pandas.pydata.org/). <br>


### 3.3 Começando com Pandas

#### 3.3.1 Como importá-lo?

Por algum motivo, todo mundo importa o Pandas da mesma forma, como mostramos abaixo. Aproveitamos também para importar o Numpy, biblioteca que já apresentamos anteriormente.

In [ ]:
import pandas as pd
import numpy as np

#### 3.3.2 Objetos do Pandas

Existem 2 principais tipos de objetos no Pandas: as [*Series*](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.html#pandas.Series) e os *DataFrames*. As *Series* são sequências de uma dimensão  de elementos (pra ser mais específicos *ndarray*), todos do mesmo tipo de dados, com rótulos/índices (*labels*). São o objeto primário do Pandas, tudo vai funcionar baseado nelas. Pra criar um objeto do tipo *Series*, podemos fazer o seguinte:

In [ ]:
s = pd.Series([1, 3, 5, np.nan, 6, 8])
s

0    1.0
1    3.0
2    5.0
3    NaN
4    6.0
5    8.0
dtype: float64

Podemos ver acima que cada elemento da *Series* tem um rótulo relacionado. Esses rótulos podem ser tanto numéricos quanto de texto! Ao final do objeto temos a informação sobre o tipo de dados da *Series*: nesse caso, números *float64*. <br>
O outro objeto principal do Pandas é o [*DataFrame*](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.html#pandas.DataFrame), que é basicamente uma coleção de *Series* com rótulos em comum. O *DataFrame* é bem parecido com uma tabela de Excel, com seus índices e colunas. Tanto suas linhas quanto suas colunas tem rótulos, nos permitindo acessar qualquer célula pela sua coordenada. Há diversas formas de criar um *DataFrame*, vamos começar com o mais simples:

In [ ]:
dates = pd.date_range('20130101', periods=6) #estamos criando uma lista de datas entre 01/01/2013 e 06/01/2013. Note que estamos usando uma função do Pandas
df = pd.DataFrame(
    np.random.randn(6, 4), #apenas números aleatórios nas células
    index=dates,  #especificando quais são os índices. Eles aceitam até datas como índice! Isso é muito bom para lidar com dados seriados no tempo
    columns=list('ABCD')) #especificando como quero que sejam os nomes das colunas passando uma lista de letras
df

A         B         C         D
2013-01-01 -0.455710  0.624624  1.892776  0.509263
2013-01-02 -0.992732 -1.268325  0.925892  1.107513
2013-01-03 -0.167113  1.325625 -1.873007 -0.931289
2013-01-04 -0.335453  1.422685  0.215810 -0.292181
2013-01-05 -1.364211  0.018419 -1.518015  0.567646
2013-01-06  0.902999 -0.879717 -1.276829 -0.175295

Isso é um *DataFrame*! Uma coisa boa do Pandas com o Jupyter Notebook é que eles mostram o *DataFrame* de uma forma bastante amigável. Agora que sabemos sobre as duas principais estruturas de dados do Pandas, podemos aprender sobre os principais métodos e funcionalidades dessa biblioteca, e para isso vamos utilizar dados reais sobre um das marcas mais conhecidas dos desenhos e videogames :)

Observação: um *DataFrame* pode ser visto como um dicionário de listas.

#### 3.3.3 Carregando o conjunto de dados

Um dos tipos de dados mais comuns para se guardar arquivos são os CSVs. O Pandas tem diversas funções para transformar os mais variados tipos de arquivos em *DataFrames*, como csv, excel, json e etc. Nesse exemplo, vamos usar o [leitor de csv](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_csv.html).

In [ ]:
pkmn = pd.read_csv(
    './dados/aula_1_pokemon.csv', #o caminho para o arquivo que se quer ler
    sep=',') #o caracter utilizado para separar os valores

#### 3.3.4 Visualizações iniciais
Ótimo! Acabamos de criar um *DataFrame* a partir de um arquivo csv. Mas como gostamos de verificar as coisas, seria interessante saber algumas coisas como o que os dados contêm, como é, se tem valores nulos e etc. O Pandas tem 4 métodos principais para isso:

In [ ]:
pkmn.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 800 entries, 0 to 799
Data columns (total 13 columns):
#             800 non-null int64
Name          800 non-null object
Type 1        800 non-null object
Type 2        414 non-null object
Total         800 non-null int64
HP            800 non-null int64
Attack        800 non-null int64
Defense       800 non-null int64
Sp. Atk       800 non-null int64
Sp. Def       800 non-null int64
Speed         800 non-null int64
Generation    800 non-null int64
Legendary     800 non-null bool
dtypes: bool(1), int64(9), object(3)
memory usage: 75.9+ KB


A primeira coisa que gosto de fazer após ler um conjunto de dados em um *DataFrame* é utilizar o método *info()*. Ele mostra informações como:
 - a classe do objeto criado
 - o intervalo do índice e quantas linhas de dados se tem
 - as colunas, seus nomes e tipos de dado
 - quais os tipos de dados presentes no *DataFrame* e quantas colunas de cada
 - a quantidade de memória utilizada pelo computador para guardar esses dados

Podemos ver que temos uma base sobre Pokémon, com 800 linhas com índices númericos de 0 a 799, 13 colunas de três tipos de dados diferentes, usando cerca de ~76kB de memória RAM. As informações presentes sobre os Pokémons são o número, o nome, o tipo (alguns tem subtipo, mas não todos, por isso dos dados faltantes na coluna *Type 2*), estatísticas de ataque, defesa e velocidade, a geração e a indicação se ele é lendário (como uma espécie mística). <br> Todas essas informações a gente conseguiu descobrir com apenas uma linha de código! Vamos então ver como o *DataFrame* realmente é. Temos dois métodos para isso:

In [ ]:
pkmn.head()

#                   Name Type 1  Type 2  Total  HP  Attack  Defense  \
0  1              Bulbasaur  Grass  Poison    318  45      49       49   
1  2                Ivysaur  Grass  Poison    405  60      62       63   
2  3               Venusaur  Grass  Poison    525  80      82       83   
3  3  VenusaurMega Venusaur  Grass  Poison    625  80     100      123   
4  4             Charmander   Fire     NaN    309  39      52       43   

   Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0       65       65     45           1      False  
1       80       80     60           1      False  
2      100      100     80           1      False  
3      122      120     80           1      False  
4       60       50     65           1      False

In [ ]:
pkmn.tail()

#                 Name   Type 1 Type 2  Total  HP  Attack  Defense  \
795  719              Diancie     Rock  Fairy    600  50     100      150   
796  719  DiancieMega Diancie     Rock  Fairy    700  50     160      110   
797  720  HoopaHoopa Confined  Psychic  Ghost    600  80     110       60   
798  720   HoopaHoopa Unbound  Psychic   Dark    680  80     160       60   
799  721            Volcanion     Fire  Water    600  80     110      120   

     Sp. Atk  Sp. Def  Speed  Generation  Legendary  
795      100      150     50           6       True  
796      160      110    110           6       True  
797      150      130     70           6       True  
798      170      130     80           6       True  
799      130       90     70           6       True

O método *.head()* e *.tail()* mostram, respectivamente, as primeiras e últimas n linhas do *DataFrame* (por padrão, n=5, mas você pode passar qualquer número como parâmetro) mostrando os índices e os nomes das colunas como numa tabela.

In [ ]:
pkmn.describe()

#      Total          HP      Attack     Defense     Sp. Atk  \
count  800.000000  800.00000  800.000000  800.000000  800.000000  800.000000   
mean   362.813750  435.10250   69.258750   79.001250   73.842500   72.820000   
std    208.343798  119.96304   25.534669   32.457366   31.183501   32.722294   
min      1.000000  180.00000    1.000000    5.000000    5.000000   10.000000   
25%    184.750000  330.00000   50.000000   55.000000   50.000000   49.750000   
50%    364.500000  450.00000   65.000000   75.000000   70.000000   65.000000   
75%    539.250000  515.00000   80.000000  100.000000   90.000000   95.000000   
max    721.000000  780.00000  255.000000  190.000000  230.000000  194.000000   

          Sp. Def       Speed  Generation  
count  800.000000  800.000000   800.00000  
mean    71.902500   68.277500     3.32375  
std     27.828916   29.060474     1.66129  
min     20.000000    5.000000     1.00000  
25%     50.000000   45.000000     2.00000  
50%     70.000000   65.000000     3.00000  
75%     90.000000   90.000000     5.00000  
max    230.000000  180.000000     6.00000

Por fim, o método *.describe()* mostra um resumo estatístico de todas as colunas numéricas. É um método bom para ter uma ideia inicial sobre o que ocorre nas colunas de uma perspectiva estatística.

#### Exercício 3.3 (9 Pontos)
*Correção Binária: O aluno recebe todos os pontos caso cumpra todos os objetivos do exercicio (independente da lógica ou estilo da programação), caso contrário, recebe zero.*

Use o arquivo de jogadores do FIFA Ultimate Team para os exercícios de Pandas. Caso você não conheça, o Ultimate Team (FUT) é um modo de jogo do FIFA onde você monta seu próprio time comprando jogadores do jogo. <br>
Substitua os \____ abaixo para ler o arquivo e siga as instruções para ter as visualizações iniciais.

In [ ]:
# leia o arquivo (tente abrir num editor de texto antes para verificar o separador)
fut_players = pd.read_csv('./dados/aula_1_fut_players.csv', sep=',')

# mostre as primeiras 10 linhas de dados
fut_players.head(10)

player_id player_name          player_extended_name      quality revision  \
0          1        Pelé      Arantes Nascimento Edson  Gold - Rare     Icon   
1          2    Maradona                Diego Maradona  Gold - Rare     Icon   
2          3     Ronaldo  Nazário de Lima Ronaldo Luís  Gold - Rare     Icon   
3          4        Pelé      Arantes Nascimento Edson  Gold - Rare     Icon   
4          5    Maradona                Diego Maradona  Gold - Rare     Icon   
5          6     Maldini                 Paolo Maldini  Gold - Rare     Icon   
6          7     Ronaldo  Nazário de Lima Ronaldo Luís  Gold - Rare     Icon   
7          8      Yashin                    Lev Yashin  Gold - Rare     Icon   
8          9  Ronaldinho      de Assis Moreira Ronaldo  Gold - Rare     Icon   
9         10  Van Basten              Marco van Basten  Gold - Rare     Icon   

   origin  overall   club league nationality  ...   cam    cf    rf    lf  \
0   Prime       98  Icons  Icons      Brazil  ...  96.0  96.0  96.0  96.0   
1   Prime       97  Icons  Icons   Argentina  ...  95.0  94.0  94.0  94.0   
2   Prime       96  Icons  Icons      Brazil  ...  91.0  94.0  94.0  94.0   
3  Medium       95  Icons  Icons      Brazil  ...  93.0  94.0  94.0  94.0   
4  Medium       95  Icons  Icons   Argentina  ...  93.0  92.0  92.0  92.0   
5   Prime       94  Icons  Icons       Italy  ...  70.0  68.0  68.0  68.0   
6  Medium       94  Icons  Icons      Brazil  ...  90.0  92.0  92.0  92.0   
7   Prime       94  Icons  Icons      Russia  ...   NaN   NaN   NaN   NaN   
8   Prime       94  Icons  Icons      Brazil  ...  93.0  92.0  92.0  92.0   
9   Prime       93  Icons  Icons     Holland  ...  86.0  90.0  90.0  90.0   

     rw    lw    st                                             traits  \
0  96.0  96.0  95.0                                       Finesse Shot   
1  94.0  94.0  90.0  Avoids Using Weaker Foot, Finesse Shot, Flair,...   
2  92.0  92.0  94.0         Tries To Beat Defensive Line, Finesse Shot   
3  94.0  94.0  92.0                                       Finesse Shot   
4  92.0  92.0  88.0  Avoids Using Weaker Foot, Finesse Shot, Flair,...   
5  69.0  69.0  70.0                                        Team Player   
6  91.0  91.0  92.0                                       Finesse Shot   
7   NaN   NaN   NaN                               Puncher, Team Player   
8  93.0  93.0  88.0                                Finesse Shot, Flair   
9  87.0  87.0  92.0                          Finesse Shot, Team Player   

                                        specialities  base_id  
0  Speedster, Aerial Threat, Dribbler, Play Maker...   237067  
1  Speedster, Dribbler, Play Maker, Distance Shoo...   190042  
2  Speedster, Dribbler, Distance Shooter, FK Spec...    37576  
3  Speedster, Dribbler, Distance Shooter, Crosser...   237068  
4  Dribbler, Play Maker, Distance Shooter, Crosse...   237074  
5  Aerial Threat, Tackler, Tactician, Complete De...   238439  
6  Speedster, Dribbler, Distance Shooter, FK Spec...   237064  
7                                                NaN   238380  
8  Speedster, Dribbler, Distance Shooter, Crosser...    28130  
9  Aerial Threat, Distance Shooter, Acrobat, Clin...   192181  

[10 rows x 82 columns]

In [ ]:
# mostre as últimas 10 linhas de dados
fut_players.tail(10)

player_id     player_name      player_extended_name        quality  \
18821      19298           Angol                 Lee Angol  Bronze - Rare   
18822      19299  Mendoza Hansen  Kevin Ray Mendoza Hansen         Bronze   
18823      19300         Piossek            Marcus Piossek         Silver   
18824      19301         Bingham            Rakish Bingham         Bronze   
18825      19302          Gibson             Jordan Gibson         Bronze   
18826      19303           Shala            Herolind Shala         Silver   
18827      19304           Boyer              Fabien Boyer         Silver   
18828      19305           Arias               Jafar Arias         Silver   
18829      19306     Moses Ekpai         Ubong Moses Ekpai         Bronze   
18830      19307          Diallo     Amadou Tidiane Diallo         Bronze   

      revision     origin  overall              club            league  \
18821   Normal  Transfers       63      Lincoln City    EFL League Two   
18822   Normal        NaN       55        AC Horsens         Superliga   
18823   Normal  Transfers       66         SV Meppen           3. Liga   
18824   Normal  Transfers       62   Cheltenham Town    EFL League Two   
18825   Normal  Transfers       51         Stevenage    EFL League Two   
18826   Normal        NaN       65          IK Start       Eliteserien   
18827   Normal        NaN       65  Grenoble Foot 38  Domino’s Ligue 2   
18828   Normal        NaN       65          FC Emmen        Eredivisie   
18829   Normal        NaN       62    Viktoria Plzeň        Česká Liga   
18830   Normal        NaN       60       Red Star FC  Domino’s Ligue 2   

       nationality  ...   cam    cf    rf    lf    rw    lw    st  \
18821      England  ...  60.0  61.0  61.0  61.0  61.0  61.0  62.0   
18822  Philippines  ...   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
18823       Poland  ...  65.0  65.0  65.0  65.0  65.0  65.0  64.0   
18824      England  ...  58.0  60.0  60.0  60.0  60.0  60.0  61.0   
18825      England  ...  48.0  50.0  50.0  50.0  51.0  51.0  50.0   
18826       Norway  ...  67.0  66.0  66.0  66.0  67.0  67.0  63.0   
18827   Madagascar  ...  46.0  46.0  46.0  46.0  48.0  48.0  48.0   
18828      Curaçao  ...  57.0  60.0  60.0  60.0  57.0  57.0  64.0   
18829      Nigeria  ...  60.0  63.0  63.0  63.0  62.0  62.0  62.0   
18830       Guinea  ...  58.0  58.0  58.0  58.0  59.0  59.0  57.0   

                             traits  specialities  base_id  
18821           Finesse Shot, Flair           NaN   210244  
18822                           NaN           NaN   215656  
18823                           NaN           NaN   197148  
18824                           NaN           NaN   209462  
18825                           NaN           NaN   240813  
18826  Speed Dribbler (CPU AI Only)           NaN   205423  
18827                  Injury Prone           NaN   208832  
18828                           NaN           NaN   226122  
18829                           NaN           NaN   244790  
18830                           NaN           NaN   233902  

[10 rows x 82 columns]

In [ ]:
# use o método .info() no DataFrame
fut_players.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18831 entries, 0 to 18830
Data columns (total 82 columns):
player_id               18831 non-null int64
player_name             18831 non-null object
player_extended_name    18831 non-null object
quality                 18831 non-null object
revision                18826 non-null object
origin                  2478 non-null object
overall                 18831 non-null int64
club                    18831 non-null object
league                  18831 non-null object
nationality             18831 non-null object
position                18831 non-null object
age                     18831 non-null int64
date_of_birth           18831 non-null object
height                  18831 non-null int64
weight                  18831 non-null int64
intl_rep                18831 non-null int64
added_date              18831 non-null object
pace                    16882 non-null float64
pace_acceleration       18831 non-null int64
pace_sprint_speed       

In [ ]:
# mostre o resumo estatístico das colunas numericas
fut_players.describe()

player_id       overall           age        height        weight  \
count  18831.000000  18831.000000  18831.000000  18831.000000  18831.000000   
mean    9724.997663     68.200839     26.364027    181.413202     75.565929   
std     5592.051018      8.143145      5.223632      6.778249      7.096437   
min        1.000000     47.000000     17.000000    155.000000     50.000000   
25%     4846.500000     63.000000     23.000000    177.000000     70.000000   
50%     9805.000000     67.000000     26.000000    182.000000     75.000000   
75%    14581.500000     73.000000     29.000000    186.000000     80.000000   
max    19307.000000     99.000000     89.000000    205.000000    110.000000   

           intl_rep          pace  pace_acceleration  pace_sprint_speed  \
count  18831.000000  16882.000000       18831.000000       18831.000000   
mean       1.172269     69.072859          65.947427          66.106580   
std        0.587855     11.909037          15.120939          14.851889   
min        0.000000     24.000000          13.000000          12.000000   
25%        1.000000     62.000000          58.000000          58.000000   
50%        1.000000     70.000000          68.000000          68.000000   
75%        1.000000     77.000000          76.000000          76.000000   
max        5.000000     99.000000          99.000000          99.000000   

          dribbling  ...            rm            lm           cam  \
count  16882.000000  ...  16882.000000  16882.000000  16882.000000   
mean      64.323540  ...     62.166746     62.166746     61.509774   
std       10.906452  ...     10.206237     10.206237     10.760257   
min       24.000000  ...     28.000000     28.000000     27.000000   
25%       58.000000  ...     56.000000     56.000000     55.000000   
50%       65.000000  ...     63.000000     63.000000     62.000000   
75%       71.000000  ...     68.000000     68.000000     68.000000   
max       99.000000  ...     97.000000     97.000000     98.000000   

                 cf            rf            lf            rw            lw  \
count  16882.000000  16882.000000  16882.000000  16882.000000  16882.000000   
mean      61.252517     61.252517     61.252517     61.511314     61.511314   
std       10.901557     10.901557     10.901557     10.866396     10.866396   
min       27.000000     27.000000     27.000000     26.000000     26.000000   
25%       54.000000     54.000000     54.000000     55.000000     55.000000   
50%       62.000000     62.000000     62.000000     62.000000     62.000000   
75%       68.000000     68.000000     68.000000     68.000000     68.000000   
max       98.000000     98.000000     98.000000     97.000000     97.000000   

                 st        base_id  
count  16882.000000   18831.000000  
mean      60.297832  209831.137114  
std       10.213289   31943.691550  
min       31.000000      16.000000  
25%       53.000000  194928.500000  
50%       60.000000  214860.000000  
75%       67.000000  232590.500000  
max       98.000000  245714.000000  

[8 rows x 66 columns]

### 3.4 Filtrando (*Filtering*) e fatiando (*slicing*) os dados

*Filtering* e *slicing* são técnicas utilizadas para isolar partes específicas do *DataFrame*, sejam linhas, colunas ou células. Isso é muito útil pois diversas vezes queremos analisar alguns dados ao invés da base inteira. O Pandas tem ferramentas próprias para isso. <br>

#### 3.4.1 *Slicing*

No Pandas existem duas principais formas de fatiar os dados, isto é, selecionar apenas uma parte de acordo com as linhas e colunas do *DataFrame*: utilizando o nome das partes ou com os métodos *.loc()* e *.iloc()*. Vamos começar pelo primeiro. Para fazer isso, imagine que queremos apenas os nomes e o poder de ataque dos Pokémons e veja o exemplo:

In [ ]:
pkmn[['Name','Attack']].head(7)

Name  Attack
0              Bulbasaur      49
1                Ivysaur      62
2               Venusaur      82
3  VenusaurMega Venusaur     100
4             Charmander      52
5             Charmeleon      64
6              Charizard      84

Duas coisas importantes aqui:
 - Juntamos o *slicing* com o método *.head()* na mesma linha, para que fosse possível ver o resultado do fatiamento. Ao usar o Pandas é possível e comum fazer esse tipo de agrupamento de operações.
 - Foram utilizadas chaves duplas [[]] no fatiamento. Ao fazer isso, estou explicitando que quero um objeto do tipo *DataFrame*. Se eu quisesse objetos do tipo *Series* usaria chaves simples. Podemos ver um exemplo disso abaixo:

In [ ]:
type(pkmn[['Name']])

pandas.core.frame.DataFrame

In [ ]:
type(pkmn['Name'])

pandas.core.series.Series

Outra forma de se obter um objeto do tipo *Series* é passando a coluna como se fosse um atributo do *DataFrame*:

In [ ]:
pkmn.Name

0                  Bulbasaur
1                    Ivysaur
2                   Venusaur
3      VenusaurMega Venusaur
4                 Charmander
               ...          
795                  Diancie
796      DiancieMega Diancie
797      HoopaHoopa Confined
798       HoopaHoopa Unbound
799                Volcanion
Name: Name, Length: 800, dtype: object

O único problema desse formato é que colunas cujo nome contém espaços não funcionarão, como é o caso das colunas *Type 1*, *Type 2*, *Sp. Atk* e *Sp. Def*. Para resolver isso, vamos renomeá-las com o método *.rename()*.

In [ ]:
pkmn.rename(
    columns={'Type 1':'Type_1', 'Type 2':'Type_2', 'Sp. Atk':'Sp_Atk','Sp. Def':'Sp_Def'}, #passando o nome antigo e novo como um dicionário
    inplace = True #algumas operações com Pandas criam uma cópia do DataFrame e não alteram o objeto em si, alteramos isso mudando o parâmetro inplace para verdadeiro
)

In [ ]:
pkmn.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 800 entries, 0 to 799
Data columns (total 13 columns):
#             800 non-null int64
Name          800 non-null object
Type_1        800 non-null object
Type_2        414 non-null object
Total         800 non-null int64
HP            800 non-null int64
Attack        800 non-null int64
Defense       800 non-null int64
Sp_Atk        800 non-null int64
Sp_Def        800 non-null int64
Speed         800 non-null int64
Generation    800 non-null int64
Legendary     800 non-null bool
dtypes: bool(1), int64(9), object(3)
memory usage: 75.9+ KB


Agora que os nomes foram trocados, podemos obter um objeto *Series* do tipo do Pokémon como a seguir:

In [ ]:
pkmn.Type_1

0        Grass
1        Grass
2        Grass
3        Grass
4         Fire
        ...   
795       Rock
796       Rock
797    Psychic
798    Psychic
799       Fire
Name: Type_1, Length: 800, dtype: object

Outra forma de selecionar partes dos dados é usando os métodos *.loc()* e *.iloc()*.<br>
Para usar a localização númerica utilizamos o *iloc*. Como você pode imaginar, a linhas e colunas são ordenadas por números inteiros sequenciais, começando do 0, como nas listas. Dessa forma, se você sabe o número da linha e da coluna, você pode usar o *iloc*. Por exemplo, se quisermos a coluna HP, que é a 6ª, poderíamos fazer o seguinte:

In [ ]:
pkmn.iloc[:,5].head()

0    45
1    60
2    80
3    80
4    39
Name: HP, dtype: int64

A sintaxe do *iloc* é como [x,y], que significa que queremos a (x+1)ª linha e (y+1)ª coluna. Se utilizarmos ':' no lugar de x ou y significa que queremos a coluna ou linha completa, respectivamente. Vamos pegar o HP do Bulbasaur, o primeiro Pokémon do nosso *DataFrame*:

In [ ]:
print("O HP do Bulbasaur é "+str(pkmn.iloc[0,5]))

O HP do Bulbasaur é 45


O método *.loc()* usa o rótulo para acessar os valores. Dessa forma, ao invés de passarmos as coordenadas numéricas, passamos o nome da linha e da coluna, como a seguir:

In [ ]:
pkmn.loc[0, :]

#                     1
Name          Bulbasaur
Type_1            Grass
Type_2           Poison
Total               318
HP                   45
Attack               49
Defense              49
Sp_Atk               65
Sp_Def               65
Speed                45
Generation            1
Legendary         False
Name: 0, dtype: object

No caso, os rótulos das linhas são iguais às suas coordenadas, por isso ficou parecido com o *iloc*. Vamos fazer o teste com as colunas também para ver a diferença. Abaixo pegaremos novamente o HP do Bulbasaur:

In [ ]:
print("O HP do Bulbasaur é "+str(pkmn.loc[0,'HP']))

O HP do Bulbasaur é 45


#### 3.4.2 Filtros (*Filtering*)

Uma vez sabendo isolar partes do *DataFrame* de acordo com a localização dos dados, podemos partir para isolar de acordo com condições, ou seja, filtrar os dados.
Para conseguir fazer isso no Pandas, fazemos o seguinte: passamos uma expressão condicional e o Pandas retorna apenas as partes que teriam a condição como verdade. Para testar isso, vamos ver a defesa média de todas os Pokémons e depois ver se os tipos 'Rock' e 'Steel' tem defesas maiores:

In [ ]:
pkmn.Defense.mean() #note que operações comuns como média (mean), mediana (median) e soma (sum) são métodos do Pandas

73.8425

In [ ]:
pkmn.loc[pkmn.Type_1=='Rock'].Defense.mean()

100.79545454545455

In [ ]:
pkmn[pkmn.Type_1=='Steel'].Defense.mean()

126.37037037037037

De fato, parece que os tipos selecionados tem média acima dos demais Pokémons. Para verificar isso, passamos a condição pkmn.Type_1=='Steel' entre chaves, o que retorna apenas as linhas de tal tipo. Com isso, selecionamos apenas a coluna de defesa e calculamos a média. <br>
Vamos ver agora os Pokémons com defesa maior que 150, cujo tipo principal não é 'Rock' nem 'Steel':

In [ ]:
pkmn[(pkmn.Defense > 150)&(pkmn.Type_1!='Rock')&(pkmn.Type_1!='Steel')]

#                   Name   Type_1   Type_2  Total   HP  Attack  \
87    80    SlowbroMega Slowbro    Water  Psychic    590   95      75   
98    91               Cloyster    Water      Ice    525   50      95   
230  213                Shuckle      Bug     Rock    505   20      10   
424  383  GroudonPrimal Groudon   Ground     Fire    770  100     180   
430  386    DeoxysDefense Forme  Psychic      NaN    600   50      70   
789  713                Avalugg      Ice      NaN    514   95     117   

     Defense  Sp_Atk  Sp_Def  Speed  Generation  Legendary  
87       180     130      80     30           1      False  
98       180      85      45     70           1      False  
230      230      10     230      5           2      False  
424      160     150      90     90           3       True  
430      160      70     160     90           3       True  
789      184      44      46     28           6      False

Podemos ver acima que é possível juntar condições com os operadores E (&) e OU (|), como vimos no inicio da aula.<br>
Vamos dizer agora que você quer apenas alguns Pokémons em específico, por exemplo Venusaur, Charizard e Blastoise. Criar uma condição para cada e uní-las com o operador & pode ser difícil, ainda mais se for uma quantidade grande de opções. Podemos facilitar isso passando uma tupla ao método *.isin()*, como abaixo:

In [ ]:
aux = ('Venusaur', 'Charizard', 'Blastoise')
pkmn[pkmn.Name.isin(aux)]

#       Name Type_1  Type_2  Total  HP  Attack  Defense  Sp_Atk  Sp_Def  \
2   3   Venusaur  Grass  Poison    525  80      82       83     100     100   
6   6  Charizard   Fire  Flying    534  78      84       78     109      85   
11  9  Blastoise  Water     NaN    530  79      83      100      85     105   

    Speed  Generation  Legendary  
2      80           1      False  
6     100           1      False  
11     78           1      False

Finalmente, podemos criar novos *DataFrames* de um já existente selecionando apenas algumas linhas ou colunas dele:

In [ ]:
offensive_stats = pkmn[['#','Name','Attack','Sp_Atk','Speed']] #selecionando apenas estatísticas ofensivas
defensive_stats = pkmn[['#','Name', 'HP','Defense','Sp_Def']] #selecionando apenas estatísticas defensivas

In [ ]:
offensive_stats.head()

#                   Name  Attack  Sp_Atk  Speed
0  1              Bulbasaur      49      65     45
1  2                Ivysaur      62      80     60
2  3               Venusaur      82     100     80
3  3  VenusaurMega Venusaur     100     122     80
4  4             Charmander      52      60     65

In [ ]:
defensive_stats.head()

#                   Name  HP  Defense  Sp_Def
0  1              Bulbasaur  45       49      65
1  2                Ivysaur  60       63      80
2  3               Venusaur  80       83     100
3  3  VenusaurMega Venusaur  80      123     120
4  4             Charmander  39       43      50

In [ ]:
fire_pkmn = pkmn[(pkmn.Type_1=='Fire')|(pkmn.Type_2=='Fire')] #filtrando apenas linhas com algumas condições
fire_pkmn.head()

#                       Name Type_1  Type_2  Total  HP  Attack  Defense  \
4  4                 Charmander   Fire     NaN    309  39      52       43   
5  5                 Charmeleon   Fire     NaN    405  58      64       58   
6  6                  Charizard   Fire  Flying    534  78      84       78   
7  6  CharizardMega Charizard X   Fire  Dragon    634  78     130      111   
8  6  CharizardMega Charizard Y   Fire  Flying    634  78     104       78   

   Sp_Atk  Sp_Def  Speed  Generation  Legendary  
4      60      50     65           1      False  
5      80      65     80           1      False  
6     109      85    100           1      False  
7     130      85    100           1      False  
8     159     115    100           1      False

In [ ]:
water_pkmn = pkmn[(pkmn.Type_1=='Water')|(pkmn.Type_2=='Water')] #filtrando apenas linhas com algumas condições
water_pkmn.head()

#                     Name Type_1 Type_2  Total  HP  Attack  Defense  \
9    7                 Squirtle  Water    NaN    314  44      48       65   
10   8                Wartortle  Water    NaN    405  59      63       80   
11   9                Blastoise  Water    NaN    530  79      83      100   
12   9  BlastoiseMega Blastoise  Water    NaN    630  79     103      120   
59  54                  Psyduck  Water    NaN    320  50      52       48   

    Sp_Atk  Sp_Def  Speed  Generation  Legendary  
9       50      64     43           1      False  
10      65      80     58           1      False  
11      85     105     78           1      False  
12     135     115     78           1      False  
59      65      50     55           1      False

#### Exercício 3.4 (18 Pontos)
*Correção Binária: O aluno recebe todos os pontos caso cumpra todos os objetivos do exercicio (independente da lógica ou estilo da programação), caso contrário, recebe zero.*

Siga as instruções e substitua os \____ para exercitar o que aprendemos.

In [ ]:
# mostre as 5 primeiras linhas das colunas player_name, position and nationality
fut_players[['player_name', 'position', 'nationality']].head()

player_name position nationality
0        Pelé      CAM      Brazil
1    Maradona      CAM   Argentina
2     Ronaldo       ST      Brazil
3        Pelé       CF      Brazil
4    Maradona      CAM   Argentina

In [ ]:
#renomeie as colunas player_id, player_name and player_extended_name para id, name and extended_name, respectivamente
fut_players.rename(
    columns={'player_id': 'id', 'player_name': 'name', 'player_extended_name': 'extended_name'},
    inplace = True
)

fut_players.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18831 entries, 0 to 18830
Data columns (total 82 columns):
id                   18831 non-null int64
name                 18831 non-null object
extended_name        18831 non-null object
quality              18831 non-null object
revision             18826 non-null object
origin               2478 non-null object
overall              18831 non-null int64
club                 18831 non-null object
league               18831 non-null object
nationality          18831 non-null object
position             18831 non-null object
age                  18831 non-null int64
date_of_birth        18831 non-null object
height               18831 non-null int64
weight               18831 non-null int64
intl_rep             18831 non-null int64
added_date           18831 non-null object
pace                 16882 non-null float64
pace_acceleration    18831 non-null int64
pace_sprint_speed    18831 non-null int64
dribbling            16882 non-null flo

In [ ]:
#imprima a coluna extended_name do 4534º jogador usando loc e iloc
print(fut_players.iloc[4533,2])
print(fut_players.loc[4533,'extended_name'])

Sandro Sirigu
Sandro Sirigu


In [ ]:
#nosso DataFrame tem muitas colunas
#crie outro DataFrame (fut_players_2) apenas com as colunas na lista abaixo
selected_columns = ['id', 'name', 'overall', 'nationality', 'position', 'pref_foot', 'base_id']

fut_players_2 = fut_players[selected_columns]

fut_players_2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18831 entries, 0 to 18830
Data columns (total 7 columns):
id             18831 non-null int64
name           18831 non-null object
overall        18831 non-null int64
nationality    18831 non-null object
position       18831 non-null object
pref_foot      18831 non-null object
base_id        18831 non-null int64
dtypes: int64(3), object(4)
memory usage: 1.0+ MB


In [ ]:
#queremos ver os melhores jogadores nascidos no Brasil (Brazil), isto é, aqueles com médio (overall) acima de 90
#mostre os 15 primeiros
aux_1 = fut_players_2[(fut_players_2.overall>90)&(fut_players_2.nationality=='Brazil')]
aux_1.head(15)

id        name  overall nationality position pref_foot  base_id
0          1        Pelé       98      Brazil      CAM     Right   237067
2          3     Ronaldo       96      Brazil       ST     Right    37576
3          4        Pelé       95      Brazil       CF     Right   237068
6          7     Ronaldo       94      Brazil       ST     Right   237064
8          9  Ronaldinho       94      Brazil       LW     Right    28130
37        38  Ronaldinho       91      Brazil      CAM     Right   238395
48        49        Pelé       91      Brazil       CF     Right   190043
51        52      Carlos       91      Brazil       LB      Left   238430
826      846   Neymar Jr       92      Brazil       LW     Right   190871
12632  13091     Rivaldo       92      Brazil       LW      Left     4231
12637  13096    Sócrates       91      Brazil      CAM     Right   190046
16491  16951   Neymar Jr       93      Brazil       LW     Right   190871
16551  17014   Neymar Jr       93      Brazil       LF     Right   190871
16923  17388   Neymar Jr       94      Brazil      CAM     Right   190871
17139  17604     Marcelo       94      Brazil       LB      Left   176676

In [ ]:
#vários jogadores bons!
#agora mostre os jogadores brasileiros que sejam canhotos (pref_foot é Left) ou que sejam goleiros (position é GK)
aux_2 = fut_players_2[(fut_players_2.pref_foot=='Left')&(fut_players_2.position=='GK')]
aux_2.head(20)

id       name  overall     nationality position pref_foot  base_id
184    186  Pinsoglio       72           Italy       GK      Left   189342
292    298    Ederson       87          Brazil       GK      Left   210257
296    302    Ederson       86          Brazil       GK      Left   210257
452    459    Früchtl       65         Germany       GK      Left   235266
506    513     Lloris       89          France       GK      Left   167948
508    515     Lloris       88          France       GK      Left   167948
560    569       Cech       82  Czech Republic       GK      Left    48940
580    591   Courtois       91         Belgium       GK      Left   192119
584    595   Courtois       90         Belgium       GK      Left   192119
624    636   Pickford       83         England       GK      Left   204935
677    694    Viviano       83           Italy       GK      Left   168354
698    715      Salin       73          France       GK      Left   150477
739    758      Meret       76           Italy       GK      Left   225116
743    762      Meret       75           Italy       GK      Left   225116
752    772   Casillas       83           Spain       GK      Left     5479
755    775   Casillas       82           Spain       GK      Left     5479
889    909       Adán       84           Spain       GK      Left   178086
890    910       Adán       83           Spain       GK      Left   178086
947    970     Foster       77         England       GK      Left   163155
1015  1039      Özcan       75         Austria       GK      Left   152211

### 3.4 Juntando DataFrames

É muito comum ter a necessidade de juntar *DataFrames* diferentes. Se você já utilizou SQL ou qualquer outro banco de dados relacional, deve conhecer isso como *join*. O Pandas também tem a mesma função utilizando o método *.merge()*. Antes do exemplo, vamos relembrar os tipos de *joins* mais comuns:<br>
![Joining Methods](https://arquivo.devmedia.com.br/artigos/Fernanda_sallai/sql_join/image001.jpg) <br>
Agora, vamos testar os *merge* nos *DataFrames* filtrados que criamos anteriormente:

In [ ]:
all_stats = pd.merge(
    offensive_stats, #o DataDrame da esquerda
    defensive_stats, #o DataDrame da direita
    how='inner', #o tipo de join que queremos fazer
    on=['#','Name']) #baseado em quais valores em comum
all_stats.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 800 entries, 0 to 799
Data columns (total 8 columns):
#          800 non-null int64
Name       800 non-null object
Attack     800 non-null int64
Sp_Atk     800 non-null int64
Speed      800 non-null int64
HP         800 non-null int64
Defense    800 non-null int64
Sp_Def     800 non-null int64
dtypes: int64(7), object(1)
memory usage: 56.2+ KB


Ótimo! Conseguimos fazer o *merge* (termo mais utilizado no Pandas) de dois *DataDrames*. Lembre-se que *inner*, *left*, *right* e *outer* terão resultados diferentes:

In [ ]:
fire_pkmn.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 64 entries, 4 to 799
Data columns (total 13 columns):
#             64 non-null int64
Name          64 non-null object
Type_1        64 non-null object
Type_2        36 non-null object
Total         64 non-null int64
HP            64 non-null int64
Attack        64 non-null int64
Defense       64 non-null int64
Sp_Atk        64 non-null int64
Sp_Def        64 non-null int64
Speed         64 non-null int64
Generation    64 non-null int64
Legendary     64 non-null bool
dtypes: bool(1), int64(9), object(3)
memory usage: 6.6+ KB


In [ ]:
water_pkmn.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 126 entries, 9 to 799
Data columns (total 13 columns):
#             126 non-null int64
Name          126 non-null object
Type_1        126 non-null object
Type_2        67 non-null object
Total         126 non-null int64
HP            126 non-null int64
Attack        126 non-null int64
Defense       126 non-null int64
Sp_Atk        126 non-null int64
Sp_Def        126 non-null int64
Speed         126 non-null int64
Generation    126 non-null int64
Legendary     126 non-null bool
dtypes: bool(1), int64(9), object(3)
memory usage: 12.9+ KB


In [ ]:
left_fire_water = pd.merge(fire_pkmn, water_pkmn, how='left', on=['#','Name'])
left_fire_water.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 64 entries, 0 to 63
Data columns (total 24 columns):
#               64 non-null int64
Name            64 non-null object
Type_1_x        64 non-null object
Type_2_x        36 non-null object
Total_x         64 non-null int64
HP_x            64 non-null int64
Attack_x        64 non-null int64
Defense_x       64 non-null int64
Sp_Atk_x        64 non-null int64
Sp_Def_x        64 non-null int64
Speed_x         64 non-null int64
Generation_x    64 non-null int64
Legendary_x     64 non-null bool
Type_1_y        1 non-null object
Type_2_y        1 non-null object
Total_y         1 non-null float64
HP_y            1 non-null float64
Attack_y        1 non-null float64
Defense_y       1 non-null float64
Sp_Atk_y        1 non-null float64
Sp_Def_y        1 non-null float64
Speed_y         1 non-null float64
Generation_y    1 non-null float64
Legendary_y     1 non-null object
dtypes: bool(1), float64(8), int64(9), object(6)
memory usage: 12.1+ KB

In [ ]:
right_fire_water = pd.merge(fire_pkmn, water_pkmn, how='right', on=['#','Name'])
right_fire_water.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 126 entries, 0 to 125
Data columns (total 24 columns):
#               126 non-null int64
Name            126 non-null object
Type_1_x        1 non-null object
Type_2_x        1 non-null object
Total_x         1 non-null float64
HP_x            1 non-null float64
Attack_x        1 non-null float64
Defense_x       1 non-null float64
Sp_Atk_x        1 non-null float64
Sp_Def_x        1 non-null float64
Speed_x         1 non-null float64
Generation_x    1 non-null float64
Legendary_x     1 non-null object
Type_1_y        126 non-null object
Type_2_y        67 non-null object
Total_y         126 non-null int64
HP_y            126 non-null int64
Attack_y        126 non-null int64
Defense_y       126 non-null int64
Sp_Atk_y        126 non-null int64
Sp_Def_y        126 non-null int64
Speed_y         126 non-null int64
Generation_y    126 non-null int64
Legendary_y     126 non-null bool
dtypes: bool(1), float64(8), int64(9), object(6)
memory u

In [ ]:
inner_fire_water = pd.merge(fire_pkmn, water_pkmn, how='inner', on=['#','Name'])
inner_fire_water.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1 entries, 0 to 0
Data columns (total 24 columns):
#               1 non-null int64
Name            1 non-null object
Type_1_x        1 non-null object
Type_2_x        1 non-null object
Total_x         1 non-null int64
HP_x            1 non-null int64
Attack_x        1 non-null int64
Defense_x       1 non-null int64
Sp_Atk_x        1 non-null int64
Sp_Def_x        1 non-null int64
Speed_x         1 non-null int64
Generation_x    1 non-null int64
Legendary_x     1 non-null bool
Type_1_y        1 non-null object
Type_2_y        1 non-null object
Total_y         1 non-null int64
HP_y            1 non-null int64
Attack_y        1 non-null int64
Defense_y       1 non-null int64
Sp_Atk_y        1 non-null int64
Sp_Def_y        1 non-null int64
Speed_y         1 non-null int64
Generation_y    1 non-null int64
Legendary_y     1 non-null bool
dtypes: bool(2), int64(17), object(5)
memory usage: 186.0+ bytes


Como podemos ver, os resultados são de fato bem diferentes.<br>
Podemos também querer apenas concatenar dois *DataDrames*, isto é, juntá-los colocando um abaixo ou ao lado do outro. Para isso, utilizamos o método *.concat()*:

In [ ]:
fire_and_water = pd.concat([fire_pkmn, water_pkmn], ignore_index=True)
fire_and_water.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 190 entries, 0 to 189
Data columns (total 13 columns):
#             190 non-null int64
Name          190 non-null object
Type_1        190 non-null object
Type_2        103 non-null object
Total         190 non-null int64
HP            190 non-null int64
Attack        190 non-null int64
Defense       190 non-null int64
Sp_Atk        190 non-null int64
Sp_Def        190 non-null int64
Speed         190 non-null int64
Generation    190 non-null int64
Legendary     190 non-null bool
dtypes: bool(1), int64(9), object(3)
memory usage: 18.1+ KB


Acima fizemos a concatenação vertical. Vamos fazer a horizontal abaixo:

In [ ]:
atk_and_defense = pd.concat([offensive_stats, defensive_stats], axis=1)
atk_and_defense.head()

#                   Name  Attack  Sp_Atk  Speed  #                   Name  \
0  1              Bulbasaur      49      65     45  1              Bulbasaur   
1  2                Ivysaur      62      80     60  2                Ivysaur   
2  3               Venusaur      82     100     80  3               Venusaur   
3  3  VenusaurMega Venusaur     100     122     80  3  VenusaurMega Venusaur   
4  4             Charmander      52      60     65  4             Charmander   

   HP  Defense  Sp_Def  
0  45       49      65  
1  60       63      80  
2  80       83     100  
3  80      123     120  
4  39       43      50

#### Exercício 3.5 (9 Pontos)
*Correção Binária: O aluno recebe todos os pontos caso cumpra todos os objetivos do exercicio (independente da lógica ou estilo da programação), caso contrário, recebe zero.*

Mais uma vez, substitua os \____ de acordo com as instruções

In [ ]:
#the_best é um DataDrame dos melhores jogadores em drible (dribbling) e chute (shooting)
the_best = fut_players[(fut_players.dribbling > 90) & (fut_players.shooting > 90)][['id', 'name', 'position', 'dribbling', 'shooting', 'overall']]

#nationalities é um DataDrame da nacionalidade dos jogadores
nationalities = fut_players[['id', 'name', 'nationality']]

#faça um merge dos dois DataDrames para obter a nacionalidade dos melhores jogadores (dica: chave é o id)
the_best_nationality = pd.merge(the_best, nationalities, how='left', on=['id','name'])
the_best_nationality.head()

id      name position  dribbling  shooting  overall nationality
0   1      Pelé      CAM       96.0      96.0       98      Brazil
1   2  Maradona      CAM       97.0      93.0       97   Argentina
2   3   Ronaldo       ST       95.0      95.0       96      Brazil
3   4      Pelé       CF       95.0      93.0       95      Brazil
4   5  Maradona      CAM       95.0      91.0       95   Argentina

### 3.5 Operações em grupo

Com Pandas nós podemos aplicar operações em grupos usando o método *.groupby()*. Ele é muito útil por ser uma forma bem simples de extrair informação de dados agregados. Para utilizá-lo, passamos as colunas nas quais queremos agrupar os dados e a operação que queremos fazer. Para exemplificar, vamos ver quantos Pokémons lendários cada geração tem:

In [ ]:
pkmn.groupby('Generation').Legendary.sum() #fazendo uma soma pois a coluna Legendary é boolean

Generation
1     6.0
2     5.0
3    18.0
4    13.0
5    15.0
6     8.0
Name: Legendary, dtype: float64

Podemos obter um relatório da média de diversas colunas para cada tipo de Pokémon:

In [ ]:
pkmn.groupby('Type_1')[['HP','Attack','Defense','Sp_Atk','Sp_Def','Speed']].mean()

HP      Attack     Defense     Sp_Atk     Sp_Def       Speed
Type_1                                                                       
Bug       56.884058   70.971014   70.724638  53.869565  64.797101   61.681159
Dark      66.806452   88.387097   70.225806  74.645161  69.516129   76.161290
Dragon    83.312500  112.125000   86.375000  96.843750  88.843750   83.031250
Electric  59.795455   69.090909   66.295455  90.022727  73.704545   84.500000
Fairy     74.117647   61.529412   65.705882  78.529412  84.705882   48.588235
Fighting  69.851852   96.777778   65.925926  53.111111  64.703704   66.074074
Fire      69.903846   84.769231   67.769231  88.980769  72.211538   74.442308
Flying    70.750000   78.750000   66.250000  94.250000  72.500000  102.500000
Ghost     64.437500   73.781250   81.187500  79.343750  76.468750   64.343750
Grass     67.271429   73.214286   70.800000  77.500000  70.428571   61.928571
Ground    73.781250   95.750000   84.843750  56.468750  62.750000   63.906250
Ice       72.000000   72.750000   71.416667  77.541667  76.291667   63.458333
Normal    77.275510   73.469388   59.846939  55.816327  63.724490   71.551020
Poison    67.250000   74.678571   68.821429  60.428571  64.392857   63.571429
Psychic   70.631579   71.456140   67.684211  98.403509  86.280702   81.491228
Rock      65.363636   92.863636  100.795455  63.340909  75.477273   55.909091
Steel     65.222222   92.703704  126.370370  67.518519  80.629630   55.259259
Water     72.062500   74.151786   72.946429  74.812500  70.517857   65.964286

Isso é realmente muito importante e extremamente utilizado com pandas pois conseguimos fazer análises dos grupos com apenas uma linha de código. Podemos perceber, por exemplo, que Pokémons do tipo *Flying* são especialistas em velocidade enquanto *Dragon* e *Fighting* são especialistas em ataque.

#### Exercício 3.6 (18 Pontos)
*Correção Binária: O aluno recebe todos os pontos caso cumpra todos os objetivos do exercicio (independente da lógica ou estilo da programação), caso contrário, recebe zero.*

Use o método *.groupby()* para descobrir qual país tem o melhor *overall* médio.

In [ ]:
#cria o DataDrame country_avg_overall, que tem o overall médio de cada país (nationality), usando groupby
country_avg_overall = fut_players.groupby('nationality')[['overall']].mean()

#usamos o método idxmax() para encontrar o maior overall médio
print("Melhor overall médio: \n", country_avg_overall.loc[country_avg_overall.idxmax()])
print("Overall médio do Brasil: ", country_avg_overall.loc["Brazil"])

Melhor overall médio: 
                     overall
nationality                
Dominican Republic     79.0
Overall médio do Brasil:  overall    75.219156
Name: Brazil, dtype: float64


### 3.6 Aplicando funções no Pandas

Com Pandas, nós temos um grande nível de controle de nossos dados, e somos capazes de transformá-los conforme queiramos. Nós podemos, até mesmo, executar funções em DataFrames e manipulá-lo como quisermos. Vamos revisitar o método info()
With Pandas, we have a deep level of control of our data, and are able to transform it as we like. We can even perform functions over the DataFrame and manipulate it as we like. Let's revisit the head() method:

In [ ]:
pkmn.head(13)

#                       Name Type_1  Type_2  Total  HP  Attack  Defense  \
0   1                  Bulbasaur  Grass  Poison    318  45      49       49   
1   2                    Ivysaur  Grass  Poison    405  60      62       63   
2   3                   Venusaur  Grass  Poison    525  80      82       83   
3   3      VenusaurMega Venusaur  Grass  Poison    625  80     100      123   
4   4                 Charmander   Fire     NaN    309  39      52       43   
5   5                 Charmeleon   Fire     NaN    405  58      64       58   
6   6                  Charizard   Fire  Flying    534  78      84       78   
7   6  CharizardMega Charizard X   Fire  Dragon    634  78     130      111   
8   6  CharizardMega Charizard Y   Fire  Flying    634  78     104       78   
9   7                   Squirtle  Water     NaN    314  44      48       65   
10  8                  Wartortle  Water     NaN    405  59      63       80   
11  9                  Blastoise  Water     NaN    530  79      83      100   
12  9    BlastoiseMega Blastoise  Water     NaN    630  79     103      120   

    Sp_Atk  Sp_Def  Speed  Generation  Legendary  
0       65      65     45           1      False  
1       80      80     60           1      False  
2      100     100     80           1      False  
3      122     120     80           1      False  
4       60      50     65           1      False  
5       80      65     80           1      False  
6      109      85    100           1      False  
7      130      85    100           1      False  
8      159     115    100           1      False  
9       50      64     43           1      False  
10      65      80     58           1      False  
11      85     105     78           1      False  
12     135     115     78           1      False

Existem algumas mega evoluções misturadas no dataset. Não seria legal se nós tivéssemos alguma flag que nos diria se um pokémon é mega ou não? E, por um acaso, será que os pokémons mega são mais poderosos?

Você deve ter percebido que evoluções mega têm um padrão em nosso DataFrame, algo como 'PokemonMega Pokemon'. Se nós tivermos esse padrão, podemos construir uma função que retorna True is este padrão for detectado:

In [ ]:
def is_it_mega(pokemon_name):
    """
    Recebe um nome de pokemon e diz se é uma mega evolução ou não
    I: string pokemon_name
    O: boolean para Mega evos
    """
    if 'Mega ' in pokemon_name: #é importante usar 'Mega ' e não 'mega', pois há um pokemon chamado called Yanmega e outro chamado Meganium que não são uma mega evolução
        return True
    else:
        return False

Vamos ver se funciona:

In [ ]:
is_it_mega('VenusaurMega Venusaur')

True

In [ ]:
is_it_mega('Squirtle')

False

Excelente! Seria ótimo se conseguíssemos aplicar essa função em todo nosso DataFrame. Para fazer isso, usaremos o método .apply(). Também criaremos uma coluna que é uma flag se o pokémon é mega:

In [ ]:
pkmn['Mega'] = pkmn.apply(
    lambda row: is_it_mega(row['Name']), #chamando uma função lambda que acabamos de construir
    axis=1 #qual direção queremos executar a função. 0 para horizontal, 1 para vertical
)

In [ ]:
pkmn.head()

#                   Name Type_1  Type_2  Total  HP  Attack  Defense  \
0  1              Bulbasaur  Grass  Poison    318  45      49       49   
1  2                Ivysaur  Grass  Poison    405  60      62       63   
2  3               Venusaur  Grass  Poison    525  80      82       83   
3  3  VenusaurMega Venusaur  Grass  Poison    625  80     100      123   
4  4             Charmander   Fire     NaN    309  39      52       43   

   Sp_Atk  Sp_Def  Speed  Generation  Legendary   Mega  
0      65      65     45           1      False  False  
1      80      80     60           1      False  False  
2     100     100     80           1      False  False  
3     122     120     80           1      False   True  
4      60      50     65           1      False  False

Agora, vamos verificar quão poderosos são os pokémons mega:

In [ ]:
pkmn.groupby('Mega').Total.mean()

Mega
False    423.457447
True     617.541667
Name: Total, dtype: float64

Uau! Eles têm quase 200 stat points a mais que pokémons normais! Evoluções mega são, sim, muito poderosos! Uma boa prática é sempre tentar manter nosso DataFrame organizado. A forma como os pokémons mega estão nomeados não é muito ótima, e nós já temos uma coluna com a flag para pokémons Mega, então, vamos atacar isso! A estrutura do nome de um pokémon mega é da seguinte forma: 'NomeMega Nome'. Portanto, se nós pegarmos o qeu vem após o caractere ' ', teremos o nome original do pokémon!

In [ ]:
pkmn.Name.nunique() #counts unique elements in a given column

800

In [ ]:
def get_original_name(s):
    """
    Recebe um nome de pokemon e retorna seu nome original
    I: s string
    O: string
    """
    return s.split(' ')[-1]

pkmn['Name'] = pkmn.Name.apply(lambda s: get_original_name(s)) #sobreescrevendo a coluna Name
pkmn.Name.nunique()

715

In [ ]:
pkmn.head()

#        Name Type_1  Type_2  Total  HP  Attack  Defense  Sp_Atk  Sp_Def  \
0  1   Bulbasaur  Grass  Poison    318  45      49       49      65      65   
1  2     Ivysaur  Grass  Poison    405  60      62       63      80      80   
2  3    Venusaur  Grass  Poison    525  80      82       83     100     100   
3  3    Venusaur  Grass  Poison    625  80     100      123     122     120   
4  4  Charmander   Fire     NaN    309  39      52       43      60      50   

   Speed  Generation  Legendary   Mega  
0     45           1      False  False  
1     60           1      False  False  
2     80           1      False  False  
3     80           1      False   True  
4     65           1      False  False

Parece que conseguimos limpar o nome deles! Agora nós cobrimos todas a parte básica de Pandas! Vamos praticar essa última parte!

#### Exercício 3.7 (27 Pontos)
*Correção Binária: O aluno recebe todos os pontos caso cumpra todos os objetivos do exercicio (independente da lógica ou estilo da programação), caso contrário, recebe zero.*

Crie uma função que retorna a classificação para o jogador de acordo com as instruções abaixo, então aplique isso para o dataframe fut_players.

In [ ]:
def get_classification(overall):
    """
    Recebe um overall de algum jogador e retorna a classificação conforme a seguir:
    Overall -> classification
    -50     -> "Amador"
    50-60   -> "Ruim"
    60-70   -> "Ok"
    70-80   -> "Bom"
    80-90   -> "Ótimo"
    90+     -> "Lenda"
    
    I: int overall
    O: string
    """
    if overall < 50:
        classificacao = 'Amador'
    elif overall >= 50 & overall < 60:
        classificacao = 'Ruim'
    elif overall >= 60 & overall < 70:
        classificacao = 'Ok'
    elif overall >= 70 & overall < 80:
        classificacao = 'Bom'
    elif overall >= 80 & overall < 90:
        classificacao = 'Ótimo'
    elif overall >= 90:
        classificacao = 'Lenda'
    else:
        classificacao = 'Fora dos limites'
        
    return classificacao
    
fut_players["classification"] = fut_players['overall'].apply(get_classification)
fut_players.head()

id      name                 extended_name      quality revision  origin  \
0   1      Pelé      Arantes Nascimento Edson  Gold - Rare     Icon   Prime   
1   2  Maradona                Diego Maradona  Gold - Rare     Icon   Prime   
2   3   Ronaldo  Nazário de Lima Ronaldo Luís  Gold - Rare     Icon   Prime   
3   4      Pelé      Arantes Nascimento Edson  Gold - Rare     Icon  Medium   
4   5  Maradona                Diego Maradona  Gold - Rare     Icon  Medium   

   overall   club league nationality  ...    cf    rf    lf    rw    lw    st  \
0       98  Icons  Icons      Brazil  ...  96.0  96.0  96.0  96.0  96.0  95.0   
1       97  Icons  Icons   Argentina  ...  94.0  94.0  94.0  94.0  94.0  90.0   
2       96  Icons  Icons      Brazil  ...  94.0  94.0  94.0  92.0  92.0  94.0   
3       95  Icons  Icons      Brazil  ...  94.0  94.0  94.0  94.0  94.0  92.0   
4       95  Icons  Icons   Argentina  ...  92.0  92.0  92.0  92.0  92.0  88.0   

                                              traits  \
0                                       Finesse Shot   
1  Avoids Using Weaker Foot, Finesse Shot, Flair,...   
2         Tries To Beat Defensive Line, Finesse Shot   
3                                       Finesse Shot   
4  Avoids Using Weaker Foot, Finesse Shot, Flair,...   

                                        specialities  base_id  classification  
0  Speedster, Aerial Threat, Dribbler, Play Maker...   237067            Ruim  
1  Speedster, Dribbler, Play Maker, Distance Shoo...   190042            Ruim  
2  Speedster, Dribbler, Distance Shooter, FK Spec...    37576            Ruim  
3  Speedster, Dribbler, Distance Shooter, Crosser...   237068            Ruim  
4  Dribbler, Play Maker, Distance Shooter, Crosse...   237074            Ruim  

[5 rows x 83 columns]

<a id="encerramento"></a>
# 4. Encerramento

Obrigado por participar da trilha, você acaba de finalizar a primeira aula. Neste momento você já deve ser capaz de manipular seus dados no Python, utilizando as bibliotecas que acabamos de aprender! 

Lembre-se que sempre que surgir alguma dúvida, você pode olhar a documentação do [Numpy](https://numpy.org/doc/) e do [Pandas](https://pandas.pydata.org/docs/).

Na próxima aula vamos aprofundar um pouco mais nas visualizações iniciais dos nossos dados, com a parte de EDA (Exploratory Data Analysis) e também veremos o primeiro tipo de algoritmo de ML (Aprendizado Não Supervisionado).